cloud/cloud_env_impl.cc

```diff
@@ -2028,7 +2034,14 @@ Status CloudEnvImpl::RollNewCookie(const std::string& local_dbname,
 
   // TODO(igor): Compact cloud manifest by looking at live files in the database
   // and removing epochs that don't contain any live files.
-  newCloudManifest->AddEpoch(delta.file_num, delta.epoch);
+  if (!newCloudManifest->AddEpoch(delta.file_num, delta.epoch)) {
+    Log(InfoLogLevel::INFO_LEVEL, info_log_,
+        "[CloudEnvImpl::RollNewCookie] delta: %s has already been applied to "
+        "current cloud "
+        "manifest",
+        delta.ToString().c_str());
+    return st;
+  }
```

Comment:
This logic is weird. In the case where the new delta is ignored we kind of half-execute this function (manifest is copied, cloud manifest isn't) and return OK?

Suggestion: Treat this case as an error. `RollNewCookie()` should always be called with a new, not an existing epoch. This is true for the callsite from `RollNewEpoch()` and it should be also true for external callsites. So, something like:

```
  auto newCloudManifest = cloud_manifest_->clone();
  std::string old_epoch = newCloudManifest->GetCurrentEpoch();
  bool update_applied = newCloudManifest->AddEpoch(delta.file_num, delta.epoch);
  if (!update_applied) {
    Log(InfoLogLevel::ERROR_LEVEL, info_log_,
      "Rolling new CLOUDMANIFEST from file number %lu, renaming MANIFEST-%s to "
      "MANIFEST-%s, new cookie: %s, error: the epoch already exists in CLOUDMANIFEST",
      delta.file_num, old_epoch.c_str(), delta.epoch.c_str(),
      cookie.c_str());
      return Status::InvalidArgument();
  }
```

Reply:
It's possible to call `RollNewCookie` with old epoch if some `new epoch records` missing in local replication log.

Agree that it's better to explicitly return non ok status in this case.

Reply:
> It's possible to call RollNewCookie with old epoch if some new epoch records missing in local replication log.

I disagree. `RollNewEpoch()` will never call `RollNewCookie()` with the old epoch, it generates a new one on every call. The external callsite (in our codebase) likewise generate a new epoch (cookie) for every call to `RollNewCookie()`. This is the follower -> presumptive leader codepath.